File: src/ios/LukuIDSDK/Tests/LukuIDSDKTests/NegativeTests.swift
// SPDX-License-Identifier: Apache-2.0
import XCTest
@testable import LukuIDSDK

final class NegativeTamperingTests: XCTestCase {
    
    private func samplesDir() -> URL {
        let start = URL(fileURLWithPath: #filePath)
        var current = start.deletingLastPathComponent()
        while true {
            let candidate = current.appendingPathComponent("samples/dotluku/dev/1.0.0")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
            let parent = current.deletingLastPathComponent()
            if parent == current { break }
            current = parent
        }
        // Fallback to searching upwards from project root if needed
        return start
    }

    func testFailsWhenAttestationSignatureIsTampered() throws {
        let samplePath = samplesDir().appendingPathComponent("first-passable-verification-sample.luku")
        guard FileManager.default.fileExists(atPath: samplePath.path) else { return }
        let data = try Data(contentsOf: samplePath)
        let archive = try LukuFile.open(data: data)
        
        // Tamper with attestation signature in the first block's first record
        if var record = archive.blocks[0].batch.first,
           let sig = record["signature"] as? String {
            var newSig = sig.replacingOccurrences(of: "A", with: "B")
            if newSig == sig {
                newSig = sig.replacingOccurrences(of: "B", with: "A")
            }
            if newSig == sig {
                newSig = sig + "X"
            }
            record["signature"] = newSig
            archive.blocks[0].batch[0] = record
        }
        
        let issues = archive.verify(options: LukuVerifyOptions(allowUntrustedRoots: false, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "RECORD_SIGNATURE_INVALID" || $0.code == "ATTESTATION_FAILED" })
    }

    func testFailsWhenAttestationDACCertIsTampered() throws {
        let samplePath = samplesDir().appendingPathComponent("first-passable-verification-sample.luku")
        guard FileManager.default.fileExists(atPath: samplePath.path) else { return }
        let data = try Data(contentsOf: samplePath)
        let archive = try LukuFile.open(data: data)
        
        // Tamper with DAC cert. Since this sample has record-level identity, we tamper with the record-level one too.
        if var record = archive.blocks[0].batch.first,
           var identity = record["identity"] as? [String: Any],
           let dac = identity["dac_der"] as? String {
            identity["dac_der"] = dac.replacingOccurrences(of: "M", with: "N")
            record["identity"] = identity
            archive.blocks[0].batch[0] = record
        }
        
        // Also tamper block level for good measure
        if let dac = archive.blocks[0].attestationDacDer {
            archive.blocks[0].attestationDacDer = dac.replacingOccurrences(of: "M", with: "N")
        }
        
        let issues = archive.verify(options: LukuVerifyOptions(allowUntrustedRoots: false, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" })
    }

    func testFailsWhenBlocksHashIsTampered() throws {
        let samplePath = samplesDir().appendingPathComponent("first-passable-verification-sample.luku")
        guard FileManager.default.fileExists(atPath: samplePath.path) else { return }
        let data = try Data(contentsOf: samplePath)
        let archive = try LukuFile.open(data: data)
        
        // Tamper with blocks_hash in manifest
        archive.manifest.blocksHash = String(repeating: "0", count: 64)
        
        let issues = archive.verify(options: LukuVerifyOptions())
        XCTAssertTrue(issues.contains { $0.code == "BLOCKS_HASH_MISMATCH" })
    }

    func testFailsWhenRootFingerprintIsTampered() throws {
        let samplePath = samplesDir().appendingPathComponent("first-passable-verification-sample.luku")
        guard FileManager.default.fileExists(atPath: samplePath.path) else { return }
        let data = try Data(contentsOf: samplePath)
        let archive = try LukuFile.open(data: data)
        
        // Tamper with root fingerprint
        if let fp = archive.blocks[0].attestationRootFingerprint {
            let index = fp.index(fp.startIndex, offsetBy: 10)
            let replacement = fp[index] == "a" ? "b" : "a"
            archive.blocks[0].attestationRootFingerprint = String(fp.prefix(10) + String(replacement) + fp.dropFirst(11))
        }
        
        let issues = archive.verify(options: LukuVerifyOptions(allowUntrustedRoots: false, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" || $0.code == "BLOCK_HASH_INVALID" })
    }

    func testFailsWithTamperedChain() throws {
        let samplePath = samplesDir().appendingPathComponent("chain-tampered.luku")
        guard FileManager.default.fileExists(atPath: samplePath.path) else { return }
        let data = try Data(contentsOf: samplePath)
        let archive = try LukuFile.open(data: data)
        
        let issues = archive.verify(options: LukuVerifyOptions(allowUntrustedRoots: false, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" || $0.code == "BLOCKS_HASH_MISMATCH" })
    }
}

File: src/rust/lukuid-sdk/tests/negative_tests.rs
// SPDX-License-Identifier: Apache-2.0
use lukuid_sdk::luku::{LukuFile, LukuVerifyOptions, LukuDeviceIdentity, LukuExportOptions};
use std::path::PathBuf;
use serde_json::{Value, json};
use ed25519_dalek::{SigningKey, Signer};
use rand_core::OsRng;
use std::collections::HashMap;

fn samples_dir() -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    while !path.join("samples").exists() {
        if !path.pop() { break; }
    }
    path.join("samples/dotluku/dev/1.0.0")
}

async fn create_valid_export(device_id: &str) -> LukuFile {
    let mut csprng = OsRng;
    let signing_key: SigningKey = SigningKey::generate(&mut csprng);
    let public_key = base64::Engine::encode(&base64::engine::general_purpose::STANDARD, signing_key.verifying_key().to_bytes());
    
    let device = LukuDeviceIdentity {
        device_id: device_id.to_string(),
        public_key: public_key.clone(),
    };

    let records = vec![
        json!({
            "type": "scan",
            "signature": base64::Engine::encode(&base64::engine::general_purpose::STANDARD, signing_key.sign(b"can1").to_bytes()),
            "previous_signature": "genesis_fake",
            "canonical_string": "can1",
            "payload": {
                "ctr": 1,
                "timestamp_utc": 1000,
                "genesis_hash": "genesis_fake"
            }
        })
    ];

    let temp_dir = std::env::temp_dir();
    let path = temp_dir.join(format!("{}.luku", device_id));
    
    LukuFile::export_with_identity(
        records,
        &path,
        device,
        HashMap::new(),
        &signing_key,
        LukuExportOptions::default()
    ).unwrap();

    let data = std::fs::read(&path).unwrap();
    let _ = std::fs::remove_file(&path);
    LukuFile::open_bytes(&data).unwrap()
}

fn test_options() -> LukuVerifyOptions {
    LukuVerifyOptions {
        allow_untrusted_roots: false,
        trust_profile: "dev".to_string(),
        ..Default::default()
    }
}

#[tokio::test]
async fn test_fails_when_attestation_signature_is_tampered() {
    let mut archive = create_valid_export("LUK-TAMPER-1").await;
    
    // Tamper with signature in the first record
    if let Some(record) = archive.blocks[0].batch.get_mut(0) {
        if let Some(sig) = record.get_mut("signature") {
            if let Some(sig_str) = sig.as_str() {
                let mut new_sig = sig_str.replace('A', "B");
                if new_sig == sig_str {
                    new_sig = sig_str.replace('B', "A");
                }
                if new_sig == sig_str {
                    new_sig = format!("{}X", sig_str);
                }
                *sig = Value::String(new_sig);
            }
        }
    }
    
    let issues = archive.verify(test_options());
    if !issues.iter().any(|i| i.code == "RECORD_SIGNATURE_INVALID" || i.code == "ATTESTATION_FAILED") {
        println!("ISSUES (sig): {:?}", issues);
    }
    assert!(issues.iter().any(|i| i.code == "RECORD_SIGNATURE_INVALID" || i.code == "ATTESTATION_FAILED"));
}

#[tokio::test]
async fn test_fails_when_batch_hash_is_tampered() {
    let mut archive = create_valid_export("LUK-TAMPER-2").await;
    
    // Tamper with the batch_hash of the first block
    archive.blocks[0].batch_hash = "0".repeat(64);
    
    let options = LukuVerifyOptions::default();
    let issues = archive.verify(options);
    assert!(issues.iter().any(|i| i.code == "BLOCK_BATCH_HASH_INVALID"));
}

#[tokio::test]
async fn test_fails_when_root_fingerprint_is_tampered() {
    let mut archive = create_valid_export("LUK-TAMPER-3").await;
    
    archive.blocks[0].attestation_root_fingerprint = Some("0".repeat(64));
    
    let issues = archive.verify(test_options());
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED" || i.code == "BLOCK_HASH_INVALID"));
}

#[tokio::test]
async fn test_fails_when_attestation_dac_cert_is_tampered() {
    let mut archive = create_valid_export("LUK-TAMPER-4").await;
    archive.blocks[0].attestation_dac_der = Some("fake_der_data".to_string());
    
    let issues = archive.verify(test_options());
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED" || i.code == "ATTESTATION_CHAIN_MISSING"));
}

#[tokio::test]
async fn test_fails_when_attestation_intermediate_cert_is_tampered() {
    let mut archive = create_valid_export("LUK-TAMPER-5").await;
    archive.blocks[0].attestation_intermediate_der = Some("fake_der_data".to_string());
    
    let issues = archive.verify(test_options());
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED" || i.code == "ATTESTATION_CHAIN_MISSING"));
}

#[tokio::test]
async fn test_fails_with_tampered_chain() {
    let sample_path = samples_dir().join("chain-tampered.luku");
    if !sample_path.exists() { return; }
    let data = std::fs::read(sample_path).unwrap();
    let archive = LukuFile::open_bytes(&data).unwrap();
    
    let issues = archive.verify(test_options());
    if !issues.iter().any(|i| i.code == "ATTESTATION_FAILED" || i.code == "BLOCKS_HASH_MISMATCH") {
        println!("ISSUES (chain): {:?}", issues);
    }
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED" || i.code == "BLOCKS_HASH_MISMATCH"));
}
